ttnn/cpp/ttnn/operations/normalization/softmax/device/softmax_op.cpp

```diff
@@ -0,0 +1,219 @@
+    std::vector<Tensor> &output_tensors
+) const {
+    auto& input_tensor = input_tensors.at(0);
+    auto& output_tensor = output_tensors.at(0);
+    const auto& mask = optional_input_tensors.at(0);
+    // bool causal_mask = mask.has_value() ? mask.value().get_legacy_shape()[-2] == mask.value().get_legacy_shape()[-1] : false;
+    bool causal_mask = this->is_causal_mask;
+
+    return std::visit(
+        [&](const auto& program_config) -> operation::ProgramWithCallbacks {
+            using ProgramConfigType = std::decay_t<decltype(program_config)>;
+            if constexpr (
+                std::is_same_v<ProgramConfigType, SoftmaxShardedMultiCoreProgramConfig>
+            ) {
+                return scale_mask_softmax_sharded_multi_core(
+                    input_tensor,
+                    output_tensor,
+                    mask,
+                    this->scale,
+                    causal_mask,
+                    this->is_scale_causal_mask_hw_dims_softmax,
+                    program_config.compute_with_storage_grid_size,
+                    program_config.subblock_w,
+                    program_config.block_h,
+                    program_config.block_w,
+                    this->compute_kernel_config);
+            }
+            else {
+                return scale_mask_softmax_multi_core(input_tensor, output_tensor, mask, this->scale, causal_mask, this->compute_kernel_config);
+            }
+        },
+        this->program_config
+    );
+}
+
+const operation::Hash Softmax::compute_program_hash(
+    const std::vector<Tensor> &input_tensors,
+    const std::vector<std::optional<const Tensor>>& optional_input_tensors) const {
+    return operation::hash_operation<Softmax>(
+        std::get<DeviceStorage>(input_tensors.at(0).storage()).memory_config(),
+        input_tensors.at(0).dtype(),
+        optional_input_tensors.at(0).has_value() ? std::optional{std::get<DeviceStorage>(optional_input_tensors.at(0).value().storage()).memory_config()}
+                                                 : std::nullopt,
+        optional_input_tensors.at(0).has_value() ? std::optional{optional_input_tensors.at(0).value().dtype()}
+                                                 : std::nullopt,
+        this->output_mem_config);
+}
+
+Tensor softmax_in_place(Tensor& input_tensor, const SoftmaxProgramConfig& program_config, std::optional<const DeviceComputeKernelConfig> compute_kernel_config) {
+    return scale_mask_softmax_in_place(input_tensor, std::nullopt, std::nullopt, program_config, false, compute_kernel_config);
+}
+
+Tensor scale_mask_softmax_in_place(Tensor& input_tensor, std::optional<float> scale, std::optional<const Tensor> mask, const SoftmaxProgramConfig& program_config, const bool is_causal_mask, std::optional<const DeviceComputeKernelConfig> compute_kernel_config) {
+    std::vector<Tensor> dummy_output_tensors = {Tensor(operation::get_workers_for_op_output({input_tensor}))};
+    operation::launch_op(
```

Comment:
This is a little unexpected. Ideally they should not be using launch_op anymore.

Reply:
what ttnn operation should do is basically this

```
auto& input_tensor = input_tensors.at(0);
auto& mask = optional_input_tensors.at(0);
auto kernel_config_val = init_device_compute_kernel_config(input_tensor.device()->arch(), compute_kernel_config, MathFidelity::HiFi4, true, false, false);

return operation::run(Softmax{.scale=scale, .inplace=true, .output_mem_config=input_tensor.memory_config(), .program_config=program_config, .is_causal_mask=is_causal_mask, .compute_kernel_config=kernel_config_val}, {input_tensor}, {mask}
```

maybe I missunderstand the change

Reply:
yeah I'll try change them to run from launch_op.